consensus/src/validation/step.rs

```diff
@@ -4,24 +4,163 @@
+use crate::commons::{ConsensusError, Database, RoundUpdate};
 use crate::config;
-use crate::contract_state::Operations;
+use crate::contract_state::{CallParams, Operations};
 use crate::execution_ctx::ExecutionCtx;
 use crate::validation::handler;
-
-use node_data::ledger::to_str;
-use node_data::message::{Message, Payload, Topics};
-use std::ops::Deref;
+use anyhow::anyhow;
+use dusk_bytes::DeserializableSlice;
+use node_data::bls::PublicKey;
+use node_data::ledger::{to_str, Block};
+use node_data::message;
+use node_data::message::{AsyncQueue, Message, Payload, Topics};
 use std::sync::Arc;
 use tokio::sync::Mutex;
-use tracing::debug;
+use tokio::task::JoinSet;
+use tracing::{debug, error, Instrument};
 
 pub struct ValidationStep<T, DB: Database> {
     timeout_millis: u64,
     handler: Arc<Mutex<handler::ValidationHandler<DB>>>,
     executor: Arc<Mutex<T>>,
 }
+
+impl<T: Operations + 'static, DB: Database> ValidationStep<T, DB> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn spawn_try_vote(
+        join_set: &mut JoinSet<()>,
+        candidate: Block,
+        ru: RoundUpdate,
+        step: u8,
+        outbound: AsyncQueue<Message>,
+        inbound: AsyncQueue<Message>,
+        executor: Arc<Mutex<T>>,
+    ) {
+        let hash = to_str(&candidate.header().hash);
+        join_set.spawn(
+            async move {
+                Self::try_vote(candidate, ru, step, outbound, inbound, executor)
+                    .await
+            }
+            .instrument(tracing::info_span!("voting", hash)),
+        );
+    }
+
+    async fn try_vote(
+        candidate: Block,
+        ru: RoundUpdate,
+        step: u8,
+        outbound: AsyncQueue<Message>,
+        inbound: AsyncQueue<Message>,
+        executor: Arc<Mutex<T>>,
+    ) {
+        let hash = candidate.header().hash;
+
+        // Call VST for non-empty blocks
+        if hash != [0u8; 32] {
+            if let Err(err) =
+                Self::call_vst(&candidate, &ru, step, executor).await
+            {
+                error!(
+                    event = "failed_vst_call",
+                    reason = format!("{:?}", err)
+                );
+                return;
+            }
+        }
+
+        let hdr = message::Header {
+            pubkey_bls: ru.pubkey_bls.clone(),
+            round: ru.round,
+            step,
+            block_hash: hash,
+            topic: Topics::Validation.into(),
+        };
+
+        let signature = hdr.sign(&ru.secret_key, ru.pubkey_bls.inner());
+
+        // Sign and construct validation message
+        let msg = message::Message::new_validation(
+            hdr,
+            message::payload::Validation { signature },
+        );
+
+        // Publish validation vote
+        debug!(event = "voting", vtype = "validation", hash = to_str(&hash));
+
+        // Publish
+        outbound.send(msg.clone()).await.unwrap_or_else(|err| {
+            error!("could not publish validation {:?}", err)
```

Comment:
just a nit. this can be written inline too

`error!("could not publish validation {err:?}")`